package.json

```diff
@@ -6,29 +6,33 @@
         "node": "^18.18.0 || ^20.9.0 || >=21.1.0"
     },
     "main": "lib/index.js",
+    "types": "types/index.d.ts",
     "files": [
         "lib/",
-        "configs/"
+        "configs/",
+        "types/index.d.ts"
     ],
     "peerDependencies": {
         "eslint": "^8.23.0 || >=9.0.0-0"
     },
     "dependencies": {
         "@eslint-community/eslint-utils": "^4.4.0",
+        "@types/eslint": "^8.56.2",
```

Comment:
I think we want to avoid adding `@types/eslint` to dependencies, partly because eslint itself is a peer dependency, partly because it can be a real pain when a bunch of different modules have slightly different versions of a type dependency like `@types/eslint`, as they won't be fully compatible with one another even though the code is.

There is actually a nice trick one can do (and which I automated in [`@hdsydsvenskan/ts-ignore-import`](https://www.npmjs.com/package/@hdsydsvenskan/ts-ignore-import) and now actually took time to release a modern variant of: https://github.com/voxpelli/ts-ignore-import):

```ts
declare const plugin =
  // @ts-ignore
  import('eslint').ESLint.Plugin;
```

When doing a `@ts-ignore` like that the type will resolve to `any` if the import fails, else it will resolve to whatever the types provide.

That way ones types can gracefully degrade in the odd case where the main project has not pulled in `@types/eslint`.

Reply:
or we can add `peerDependencies: { "@types/eslint": ">=8"}` (same as eslint)?

Reply:
technically, it should be optional peer-dependencies, I don't know if it's supported tho. :)

Reply:
You can make peer dependencies optional by using [`peerDependenciesMeta`](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#peerdependenciesmeta):

```json
  "peerDependencies": {
    "@types/eslint": ">=8"
  },
  "peerDependenciesMeta": {
    "@types/eslint": {
      "optional": true
    }
  }
```

But I think its kind of redundant? If people wants types then they should install it themselves.

---

Speaking of [`ts-ignore-import`](https://github.com/voxpelli/ts-ignore-import), I used it in my latest module I released right now: https://github.com/voxpelli/list-dependents It currently ignores `@types/hast` and `read-pkg` in the types

Reply:
I think I want to make an issue for this, so we can sort this afterwards! This PR is already a little larger than I like to make PRs :sweat_smile:

I can make the ticket if you agree :)

Reply:
`@ts-ignore` works for me. I agreed it won't be a blocker to merge this PR. 😄

Reply:
Issue made here: https://github.com/eslint-community/eslint-plugin-n/issues/213
Will mark this as resolved for now.